File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int ans = 0;
        stack<int>next_min;
        stack<int>prev_min;
        heights.push_back(0);
        int n = heights.size();
        vector<int>prev(n);

        for(int i=n-1;i>=0;i--){
            while(!prev_min.empty() and heights[prev_min.top()]>heights[i]){
                prev[prev_min.top()]=i;
                prev_min.pop();
            }
            prev_min.push(i);
        }
        while(!prev_min.empty()){
            prev[prev_min.top()]=-1;
            prev_min.pop();
        }
        // cout<<prev[3]<<' '<<prev[4]<<endl;
        for(int i=0;i<n;i++){
             while(!next_min.empty() and heights[next_min.top()]>heights[i]){
                int width = i-prev[next_min.top()]-1;
                ans = max(ans,heights[next_min.top()]*width);
                next_min.pop();
            }
            next_min.push(i);
        }

        
        
        return ans;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-0.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n=nums.length;
        int[] dp = new int[n+1];
        if(n==1) return 1;
        dp[0]=1;
        if(nums[1]>nums[0]) dp[1]=2;
        else dp[1]=1;
        int ans = Math.max(dp[0],dp[1]);
        for(int i=2;i<n;i++){
            int curr = 1;
            for(int j=0;j<i;j++){
                if(nums[j]<nums[i]){
                    curr = Math.max(curr,1+dp[j]);
                }

            }
            dp[i]=curr;
            ans = Math.max(curr,ans);
        }
        return ans;
    
        
    }
}
